File: algorithm/replace-space.go
package main

import "fmt"

// 遍历添加
func replaceSpace1(s string) string {
	b := []byte(s)
	result := make([]byte, 0)
	for i := 0; i < len(b); i++ {
		if b[i] == ' ' {
			result = append(result, []byte("%20")...)
		} else {
			result = append(result, b[i])
		}
	}
	return string(result)
}

// func replaceSpace2(s string) string {
// 	bytes := []byte(s)
// 	l := len(bytes)
// 	count := 0
// 	for i := 0; i < l; i++ {
// 		if bytes[i] == ' ' {
// 			count++
// 		}
// 	}
// 	fmt.Println("count", count)
// 	result := make([]byte, l+count*2)
// 	fmt.Printf("cap: %v len: %v\n", cap(result), len(result))
// 	j := 0
// 	for i := 0; i < l; i++ {
// 		if bytes[i] == ' ' {
// 			result[j] = '%'
// 			result[j+1] = '2'
// 			result[j+2] = '0'
// 			j += 2
// 		} else {
// 			result[j] = bytes[j]
// 			j++
// 		}
// 	}
// 	fmt.Println("result", result)

// 	return string(result)
// }

// 原地修改
func replaceSpace(s string) string {
	b := []byte(s)
	length := len(b)
	spaceCount := 0
	// 计算空格数量
	for _, v := range b {
		if v == ' ' {
			spaceCount++
		}
	}
	// 扩展原有切片
	resizeCount := spaceCount * 2
	tmp := make([]byte, resizeCount)
	b = append(b, tmp...)

	i := length - 1
	j := len(b) - 1
	for i >= 0 {
		if b[i] != ' ' {
			b[j] = b[i]
			i--
			j--
		} else {
			b[j] = '0'
			b[j-1] = '2'
			b[j-2] = '%'
			i--
			j -= 3
		}
	}

	return string(b)
}

func main() {
	str := replaceSpace("We are happy.")
	fmt.Println(str)
	fmt.Println(len([]byte(" ")))
}

File: algorithm/string/replace-space.rb
=begin
题目：剑指Offer 05.替换空格
力扣题目链接(https://leetcode-cn.com/problems/ti-huan-kong-ge-lcof/)

请实现一个函数，把字符串 s 中的每个空格替换成"%20"。

示例 1
输入：s = "We are happy."
输出："We%20are%20happy."
=end

# 遍历添加
def replace_space1(s)
	res = ''
	s.each_char do |c|
		if c == ' '
			res += '%20'
		else
			res += c
		end
	end
	
	res
end

# 原地修改
def replace_space(s)
	chars = s.chars
	length = chars.length
	# 计算空格数量
	space_count = chars.count{|c| c == ' '}
	
	# 扩展原有数组
	resize_count = space_count * 2
	chars += Array.new(resize_count, ' ')

	i = length - 1
	j = chars.length - 1
	while i >= 0
		if chars[i] != ' '
			chars[j] = chars[i]
			i -= 1
			j -= 1
		else
			chars[j] = '0'
			chars[j-1] = '2'
			chars[j-2] = '%'
			i -= 1
			j -= 3
		end
	end

	chars.join('')
end

require "minitest/autorun"

describe "replace space" do
	it "must" do
		# 输入：s = "We are happy."
		# 输出："We%20are%20happy."
		input = "We are happy."
		output = "We%20are%20happy."
		
		str = replace_space(input)
		assert_equal str, output

		str = replace_space1(input)
		assert_equal str, output
	end
end